libevmasm/RuleList.h

```diff
@@ -374,6 +390,30 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 		false
 	});
 
+	rules.push_back({
+		// AND(A, SHR(B, X)) -> SHR(B, X) if 2^B-1 == A
```

Comment:
First of all: is this comment correct?

``AND(3, SHR(2, X)) -> SHR(2,X)```is wrong, even though ``2^2-1 == 3``...

Easiest thing to write would be:

``AND(A, SHR(B, X)) -> SHR(B, X) if A = ((2^256-1) >> B)``

That would mean ``A.d() == u256(-1) >> B.d()`` below...

Reply:
Ah right, that is even simpler!

Reply:
But actually `A.d() & (u256(-1) >> B.d()) == A.d()` would catch more cases, wouldn't it? Like `0xffff & (x >> (256-8))`

Reply:
Ah yes, and A can even have higher bits set... one second...

Reply:
The comment should already be:
``// AND(A, SHR(B, X)) -> SHR(B,X) if A & ((2^256 - 1) >> B) == ((2^256 - 1) >> B)``
which would mean
`` return A.d() & (u256(-1) >> B.d()) == (u256(-1) >> B.d());``
below. That would even work for ``B >= 256``, for which it reduces to ``return A.d() & 0 == 0`` which is ``return true``, which is correct, since in this case ``SHR(B,X)`` will always be zero.

Reply:
And to confirm: It also works for ``B = 0`` - then it's ``return A.d() & u256(-1) == u256(-1)``, which means ``A`` has all bits set, making the ``AND`` redundant as well.

Reply:
wrong comment indeed

Reply:
By the way we could add/should the analog rule for ``SHL``:

``AND(A, SHL(B, X)) -> SHL(B,X) if A & ((2^256 - 1) << B) == ((2^256 - 1) << B)``